x/expiration/keeper/keeper.go

```diff
@@ -0,0 +1,565 @@
+	}
+
+	// validate module asset id
+	if _, err := sdk.AccAddressFromBech32(expiration.ModuleAssetId); err != nil {
+		// check if we're dealing with a MetadataAddress
+		if _, err2 := metadatatypes.MetadataAddressFromBech32(expiration.ModuleAssetId); err2 != nil {
+			return sdkerrors.ErrInvalidAddress.Wrapf("invalid module asset id: %v", err)
+		}
+	}
+
+	// validate block time is in the future
+	if expiration.Time.Before(ctx.BlockTime()) {
+		return types.ErrExpirationTime.Wrapf("expiration time %s must be in the future", expiration.Time)
+	}
+
+	// validate deposit
+	if err := expiration.Deposit.Validate(); err != nil {
+		return sdkerrors.ErrInvalidCoins.Wrap(err.Error())
+	}
+	deposit := expiration.Deposit
+	defaultDeposit := types.DefaultDeposit
+	if deposit.IsZero() {
+		return types.ErrInvalidDeposit.Wrap("deposit amount cannot be zero")
+	}
+	if deposit.IsAllLT(sdk.NewCoins(defaultDeposit)) {
+		return types.ErrInvalidDeposit.Wrapf("deposit amount %s is less than minimum deposit amount %s",
+			deposit.String(), defaultDeposit.String())
+	}
+
+	// validate msg
+	var msg sdk.Msg
+	if err := k.cdc.UnpackAny(&expiration.Message, &msg); err != nil {
+		return types.ErrInvalidMessage.Wrapf("failed to unpack msg: %v", err)
+	}
+	if err := msg.ValidateBasic(); err != nil {
+		return types.ErrInvalidMessage.Wrapf("failed ValidateBasic: %v", err)
+	}
+
+	// validate signers
+	if err := k.validateSigners(ctx, expiration.Owner, signers, msgTypeURL); err != nil {
+		return types.ErrInvalidSigners.Wrap(err.Error())
+	}
+
+	return nil
+}
+
+// ValidateInvokeExpiration validates an expiration record.
+func (k Keeper) ValidateInvokeExpiration(
+	ctx sdk.Context,
+	moduleAssetID string,
+	signers []string,
+	msgTypeURL string,
+) (*types.Expiration, error) {
+	expiration, err := k.GetExpiration(ctx, moduleAssetID)
+	if err != nil {
+		return expiration, err
+	}
+	if expiration == nil {
+		return expiration, types.ErrNotFound.Wrapf("module asset id [%s]", moduleAssetID)
+	}
+
+	// validate before routing as things may have
+	// changed from the time this was set to now
+	var msg sdk.Msg
+	if err := k.cdc.UnpackAny(&expiration.Message, &msg); err != nil {
+		return nil, types.ErrInvalidMessage.Wrapf("failed to unpack msg: %v", err)
+	}
+	if err := msg.ValidateBasic(); err != nil {
+		return nil, types.ErrInvalidMessage.Wrapf("failed ValidateBasic: %v", err)
+	}
+
+	// anyone can delete an expired expiration
+	if ctx.BlockTime().After(expiration.Time) {
+		return expiration, nil
+	}
+
+	// validate signers and owner
+	if err := k.validateSigners(ctx, expiration.Owner, signers, msgTypeURL); err != nil {
+		return expiration, types.ErrInvalidSigners.Wrap(err.Error())
+	}
+
+	return expiration, nil
+}
+
+// ResolveDepositor resolves the account address where the expiration message deposit will be refunded.
+func (k Keeper) ResolveDepositor(
+	ctx sdk.Context,
+	expiration types.Expiration,
+	msg *types.MsgInvokeExpirationRequest,
+) (sdk.AccAddress, error) {
```

Comment:
I feel like this is significantly more complicated than it needs to be. The account that gets the deposit should be the one signing the `MsgInvokeExpirationRequest`.

No matter what, it should be the signer of the `MsgInvokeExpirationRequest`.

If it's expired, there's nothing left to check. If it's not expired, that signer needs to be the owner.

This function would also be better named something like `GetRefundTo` since it's not actually trying to look up who made the deposit.